### PR TITLE
Fashion landmark detection

### PR DESCRIPTION
**Pull request by jin-s13**

### CODE REVIEW

File: mmpose/datasets/datasets/fashion/deepfashion_full_dataset.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+            ann_file, img_prefix, data_cfg, pipeline, test_mode=test_mode)
+
+        assert self.ann_info['num_joints'] == 8
+        self.ann_info['flip_pairs'] = [[0, 1], [2, 3], [4, 5], [6, 7]]
+
+        self.ann_info['use_different_joint_weights'] = False
+        self.ann_info['joint_weights'] = \
+            np.ones((self.ann_info['num_joints'], 1), dtype=np.float32)
+
+        self.dataset_name = 'deepfashion_upper'
```
Comment (innerlee) [inline mmpose/datasets/datasets/fashion/deepfashion_full_dataset.py:59]: upper or full?

File: mmpose/datasets/datasets/fashion/fashion_base_dataset.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+            id2name[image_id] = file_name
+            name2id[file_name] = image_id
+
+        return id2name, name2id
+
+    def _xywh2cs(self, x, y, w, h, padding=1.25):
+        """This encodes bbox(x,y,w,w) into (center, scale)
+
+        Args:
+            x, y, w, h
```
Comment (innerlee) [inline mmpose/datasets/datasets/fashion/fashion_base_dataset.py:90]: Forgot padding

File: mmpose/datasets/datasets/fashion/deepfashion_full_dataset.py (hunk 1 above)
Comment (innerlee) [reply mmpose/datasets/datasets/fashion/deepfashion_full_dataset.py:59]: A signal that we need test dataset names

File: mmpose/datasets/datasets/fashion/deepfashion_full_dataset.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                    continue
+                joints_3d = np.zeros((num_joints, 3), dtype=np.float32)
+                joints_3d_visible = np.zeros((num_joints, 3), dtype=np.float32)
+
+                keypoints = np.array(obj['keypoints']).reshape(-1, 3)
+                joints_3d[:, :2] = keypoints[:, :2]
+                joints_3d_visible[:, :2] = np.minimum(1, keypoints[:, 2:3])
+
+                # the ori image is 224x224
+                center, scale = self._xywh2cs(0, 0, 224, 224, 0.8)
```
Comment (innerlee) [inline mmpose/datasets/datasets/fashion/deepfashion_full_dataset.py:86]: Double check `0.8`

File: mmpose/datasets/datasets/fashion/deepfashion_lower_dataset.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+            ann_file, img_prefix, data_cfg, pipeline, test_mode=test_mode)
+
+        assert self.ann_info['num_joints'] == 4
+        self.ann_info['flip_pairs'] = [[0, 1], [2, 3]]
+
+        self.ann_info['use_different_joint_weights'] = False
+        self.ann_info['joint_weights'] = \
+            np.ones((self.ann_info['num_joints'], 1), dtype=np.float32)
+
+        self.dataset_name = 'deepfashion_upper'
```
Comment (innerlee) [inline mmpose/datasets/datasets/fashion/deepfashion_lower_dataset.py:55]: name

File: mmpose/datasets/datasets/fashion/deepfashion_full_dataset.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+                    'joints_3d_visible': joints_3d_visible,
+                    'dataset': self.dataset_name,
+                    'bbox': obj['bbox'],
+                    'bbox_score': 1
+                })
+            gt_db.extend(rec)
+        return gt_db
+
+    def evaluate(self, outputs, res_folder, metric='PCK', **kwargs):
+        """Evaluate freihand keypoint results. The pose prediction results will
```
Comment (innerlee) [inline mmpose/datasets/datasets/fashion/deepfashion_full_dataset.py:105]: assert that kwargs be empty?

File: mmpose/datasets/datasets/fashion/fashion_base_dataset.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+        if w > aspect_ratio * h:
+            h = w * 1.0 / aspect_ratio
+        elif w < aspect_ratio * h:
+            w = h * aspect_ratio
+
+        # pixel std is 200.0
+        scale = np.array([w / 200.0, h / 200.0], dtype=np.float32)
+
+        scale = scale * padding
```
Comment (innerlee) [inline mmpose/datasets/datasets/fashion/fashion_base_dataset.py:111]: I'm confused by the name `padding`. Could you explain it briefly?

File: mmpose/datasets/datasets/fashion/fashion_base_dataset.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+        if (not self.test_mode) and np.random.rand() < 0.3:
+            center += 0.4 * (np.random.rand(2) - 0.5) * [w, h]
+
+        if w > aspect_ratio * h:
+            h = w * 1.0 / aspect_ratio
+        elif w < aspect_ratio * h:
+            w = h * aspect_ratio
+
+        # pixel std is 200.0
```
Comment (innerlee) [inline mmpose/datasets/datasets/fashion/fashion_base_dataset.py:108]: What is `pixel std`? The std in color space (B, G, R) or in spaital (x, y)?

File: mmpose/datasets/datasets/fashion/fashion_base_dataset.py [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+        Args:
+            res_file (str): Json file stored prediction results.
+            metrics (str | list[str]): Metric to be performed.
+                Options: 'PCK', 'PCKh', 'AUC', 'EPE'.
+            pck_thr (float): PCK threshold, default as 0.2.
+            pckh_thr (float): PCKh threshold, default as 0.7.
+            auc_nor (float): AUC normalization factor, default as 30 pixel.
+
+        Returns:
+            dict: Evaluation results for evaluation metric.
```
Comment (innerlee) [inline mmpose/datasets/datasets/fashion/fashion_base_dataset.py:149]: Is info_str a dict?

File: mmpose/datasets/datasets/fashion/deepfashion_upper_dataset.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+                    'joints_3d_visible': joints_3d_visible,
+                    'dataset': self.dataset_name,
+                    'bbox': obj['bbox'],
+                    'bbox_score': 1
+                })
+            gt_db.extend(rec)
+        return gt_db
+
+    def evaluate(self, outputs, res_folder, metric='PCK', **kwargs):
+        """Evaluate freihand keypoint results. The pose prediction results will
```
Comment (innerlee) [inline mmpose/datasets/datasets/fashion/deepfashion_upper_dataset.py:103]: add assertion for kwargs

File: mmpose/datasets/datasets/fashion/deepfashion_lower_dataset.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    `Fashion Landmark Detection in the Wild' ECCV'2016
+
+    The dataset loads raw features and apply specified transforms
+    to return a dict containing the image tensors and other information.
+
+    Fashion landmark indexes for lower-body clothes::
+
+        0: 'left waistline',
+        1: 'right waistline',
+        2: 'left hem',
```
Comment (innerlee) [inline mmpose/datasets/datasets/fashion/deepfashion_lower_dataset.py:25]: does hem belongs to both upper and lower?

File: mmpose/datasets/datasets/fashion/deepfashion_full_dataset.py (hunk 3 above)
Comment (jin-s13) [reply mmpose/datasets/datasets/fashion/deepfashion_full_dataset.py]: That's a mistake.

File: mmpose/datasets/datasets/fashion/fashion_base_dataset.py (hunk 8 above)
Comment (jin-s13) [reply mmpose/datasets/datasets/fashion/fashion_base_dataset.py:149]: the returned `name_value` is a dict. `info_str` is a list.

File: mmpose/datasets/datasets/fashion/deepfashion_dataset.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+        4: 'left waistline',
+        5: 'right waistline',
+        6: 'left hem',
+        7: 'right hem'
+
+    Args:
+        ann_file (str): Path to the annotation file.
+        img_prefix (str): Path to a directory where images are held.
+            Default: None.
+        data_cfg (dict): config
```
Comment (innerlee) [inline mmpose/datasets/datasets/fashion/deepfashion_dataset.py:54]: docstring for subset

File: mmpose/datasets/datasets/fashion/deepfashion_lower_dataset.py (hunk 10 above)
Comment (jin-s13) [reply mmpose/datasets/datasets/fashion/deepfashion_lower_dataset.py]: Yes, but not sure why.

Comment (jin-s13) [reply mmpose/datasets/datasets/fashion/deepfashion_lower_dataset.py]: two types of hem maybe

File: mmpose/datasets/datasets/fashion/fashion_base_dataset.py (hunk 8 above)
Comment (innerlee) [reply mmpose/datasets/datasets/fashion/fashion_base_dataset.py:149]: But this function returns info_str

Comment (jin-s13) [reply mmpose/datasets/datasets/fashion/fashion_base_dataset.py:150]: aha.. I will fix it.

**Review by innerlee — APPROVED** [review]

File: docs/tasks/2d_fashion_landmark.md [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+It is recommended to symlink the dataset root to `$MMPOSE/data`.
+If your folder structure is different, you may need to change the corresponding paths in config files.
+
+MMPose supported datasets:
+
+- [DeepFashion](#deepfashion) \[ [Homepage](http://mmlab.ie.cuhk.edu.hk/projects/DeepFashion/LandmarkDetection.html) \]
+
+## DeepFashion (Fashion Landmark Detection, FLD)
+
+For [DeepFashion](http://mmlab.ie.cuhk.edu.hk/projects/DeepFashion/LandmarkDetection.html) datatset, images can be downloaded from [download](http://mmlab.ie.cuhk.edu.hk/projects/DeepFashion/LandmarkDetection.html).
```
Comment (innerlee) [inline docs/tasks/2d_fashion_landmark.md:12]: missing bib item

Comment (jin-s13) [reply docs/tasks/2d_fashion_landmark.md:36]: added

**Pull request merged**